plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -1152,6 +1152,8 @@ private static String escapeNamePattern(String name, String escape)
         name = name.replace(escape, escape + escape);
         name = name.replace("_", escape + "_");
         name = name.replace("%", escape + "%");
+        name = name.replace("[", escape + "[");
+        name = name.replace("]", escape + "]");
```

Comment:
Square brackets are not treated differently for SQL LIKE clause in other databases except SQL Server.

This change should go into SQLServerClient (by adding an extension point to allow subclasses to provide some list of replacements - which will be appended to default list?)

Reply:
SQL Server docs on other ones are here - https://learn.microsoft.com/en-us/sql/t-sql/language-elements/like-transact-sql?redirectedfrom=MSDN&view=sql-server-ver16

Seems like we need to escape also `'` for SQLServer from the list in the link.

Reply:
> some list of replacements - which will be appended to default list?

That would be hard to reason about.

`escapeNamePattern` _could_ become an extension point, but it would need to get a new, better name.

Reply:
Don't think so, it's not connected to SQL server, it's call to 
`ResultSet getColumns(String catalog, String schemaPattern,`
                         `String tableNamePattern, String columnNamePattern)`
        `throws SQLException;`

So it's kind of common stuff, the problem is that it's not strict table name, it's pattern, so brackets used like pattern expression [A-z] not like literal symbol, so we need to quote it, if we want to have it in table name literally

Reply:
The `DatabaseMetaData` has different implementations coming from different JDBC drivers. So while the method we use to get columns is "common" their implementation is not.

Brackets are literal in almost every database except SQL Server and Sybase from my searches so far. PostgreSQL has a different `SIMILAR TO` (in addition to LIKE) but we don't use that.

Reply:
Ok now I see, that it's specific to SQL Server, I thought at first that it will be done with Java pattern matching for all Databases in the same way, so pls ignore my previous comment

Reply:
@findepi 



>>  some list of replacements - which will be appended to default list?
> That would be hard to reason about.

I was thinking of something similar to how we add connector expression rewrites - we have some standard ones present everywhere (ANSI SQL things in this case) and then whatever extensions are needed are added by each connector.

Reply:
@hashhar in connector expression rewrites  it's a must have, since the set of supported things is clearly extensible and growing over time

but note -- even there, we don't have "default rules" being provided by BaseJdbcClient and some more by subclass; the subclass is in full control of what's going on

in here, our challenge is much simpler. Escape value for `DatabaseMetaData` call.
We should make it as readable as possible.
So base jdbc should be doing what it does today; SQL client should override (copy + extend)

Reply:
So our plan is to make `escapeNamePattern` method non-static and override it in datasource specific clients with adding additional escaping rules for specific implementation ?

Reply:
yes, i feel like that's the way to go. However, we also need to find a good, self-descriptive name for this method.
Current name can only be understood by inspecting the call sites.

Reply:
i think it's important to scope this to `DatabaseMetaData` calls.

- `DatabaseMetaData` defines how the values should be processed by a remote database, hence the default implementation here is "generally correct"
- a database (or its driver) may choose to deviate from `DatabaseMetaData` spec (free as in freedom...), but we *hope* it does so consistently across all `DatabaseMetaData` methods

is it used for table names only? I think it's used for schema names too, or column names. (And not catalog names)

`escapeObjectNameForMetadataQuery` ?

Reply:
marking as resolved since we use a good name + have this scoped to only SQLServer Client.